redfish-core/lib/power_supply_metrics.hpp

```diff
@@ -0,0 +1,370 @@
+               const std::string& chassisID, const std::string& powerSupplyID) {
+                auto getChassisID = [asyncResp, chassisID, powerSupplyID](
+                                        const std::optional<std::string>&
+                                            validChassisID) {
+                    if (!validChassisID)
+                    {
+                        BMCWEB_LOG_ERROR << "Not a valid chassis ID:"
+                                         << chassisID;
+                        messages::resourceNotFound(asyncResp->res, "Chassis",
+                                                   chassisID);
+                        return;
+                    }
+
+                    BMCWEB_LOG_DEBUG << "ChassisID: " << chassisID;
+
+                    auto getPowerSupplyHandler =
+                        [asyncResp, chassisID,
+                         powerSupplyID](const std::optional<std::string>&
+                                            validPowerSupplyPath,
+                                        [[maybe_unused]] const std::string&
+                                            validPowerSupplyService) {
+                            if (!validPowerSupplyPath)
+                            {
+                                BMCWEB_LOG_ERROR
+                                    << "Not a valid power supply ID:"
+                                    << powerSupplyID;
+                                messages::resourceNotFound(asyncResp->res,
+                                                           "PowerSupply",
+                                                           powerSupplyID);
+                                return;
+                            }
+
+                            BMCWEB_LOG_DEBUG << "PowerSupplyID: "
+                                             << powerSupplyID;
+
+                            asyncResp->res.jsonValue["@odata.type"] =
+                                "#PowerSupplyMetrics.v1_0_0.PowerSupplyMetrics";
+                            asyncResp->res.jsonValue["@odata.id"] =
+                                "/redfish/v1/Chassis/" + chassisID +
+                                "/PowerSubsystem/PowerSupplies/" +
+                                powerSupplyID + "/Metrics";
+                            asyncResp->res.jsonValue["Name"] =
+                                "Metrics for " + powerSupplyID;
+                            asyncResp->res.jsonValue["Id"] = "Metrics";
+
+                            asyncResp->res.jsonValue["Oem"]["@odata.type"] =
+                                "#OemPowerSupplyMetrics.Oem";
+                            asyncResp->res
+                                .jsonValue["Oem"]["IBM"]["@odata.type"] =
+                                "#OemPowerSupplyMetrics.IBM";
+                            getValues(asyncResp, chassisID, powerSupplyID);
+                        };
+                    redfish::power_supply_utils::getValidPowerSupplyID(
```

Comment:
Cool! This should handle 404s. E.g. /redfish/v1/Chassis/<str>/PowerSubsystem/PowerSupplies/badPowerSupply/Metrics (something I see a lot of developers struggle with)